README
                     _   _
 _ __  ___ ___      | |_| |___
| '_ \/ __/ __|_____| __| / __|
| | | \__ \__ \_____| |_| \__ \
|_| |_|___/___/      \__|_|___/

Motivation
==========

Unlike most web browser traffic, which is encrypted thanks to HTTPS, the
resolving of domain names to internet addresses still happens through DNS, an
old, unencrypted protocol. This benefits analytics companies, advertisers,
internet providers and attackers, but not the end-user, who seeks online
privacy and security.

Overview
========

nss-tls is an alternative name resolving library for Linux
(http://www.kernel.org/) distributions with glibc
(https://www.gnu.org/software/libc/).

The glibc name resolver can be configured through nsswitch.conf(5) to use
nss-tls instead of the DNS resolver, or fall back to DNS when nss-tls fails.

This way, all applications that use the standard resolver API (getaddrinfo(),
gethostbyname(), etc'), are transparently migrated from DNS to encrypted means
of name resolving, with zero application-side changes and minimal resource
consumption footprint. However, nss-tls does not deal with applications that use
their own, built-in DNS resolver.

Architecture
============

nss-tls consists of three parts:
  - nss-tlsd runs in the background and receives DNS resolving requests over a
    Unix socket.
  - libnss_tls.so is a tiny client library which delegates the resolving work to
    nss-tlsd through the Unix socket and passes the results back to the
    application. This way, applications that take advantage of nss-tls are not
    affected by the complexity and the resource consumption of the libraries it
    depends on, or the constraints they impose on applications that use them.
  - tlslookup is equivalent to nslookup(1), but uses libnss_tls.so instead of
    DNS.

nss-tls is designed with security and privacy in mind. Therefore, an
unprivileged user can start a private, unprivileged instance of nss-tlsd and
libnss-tls.so will automatically use that one, instead of the system-wide
instance of nss-tlsd. Users who don't have such a private instance will continue
to use the system-wide instance, which drops its privileges to greatly reduce
its attack surface.

Dependencies
============

nss-tls depends on:
  - glibc (https://www.gnu.org/software/libc/)
  - GLib (https://wiki.gnome.org/Projects/GLib)
  - libsoup (https://wiki.gnome.org/Projects/libsoup)
  - JSON-GLib (https://wiki.gnome.org/Projects/JsonGlib)

If systemd (https://www.freedesktop.org/wiki/Software/systemd/) is present,
the installation of nss-tls includes unit files for nss-tlsd.

However, nss-tlsd does not depend on systemd. When systemd is not present, other
means of running a nss-tlsd instance for each user (e.g. xinitrc) and root (e.g.
an init script) should be used.

nss-tls uses Meson (http://mesonbuild.com/) as its build system.

On Debian (http://www.debian.org/) and derivatives, these dependencies can be
obtained using:

  1) apt install libglib2.0-dev libsoup2.4-dev libjson-glib-dev ninja-build
                 python3-pip
  2) pip3 install meson

Usage
=====

Assuming your system runs systemd
(https://www.freedesktop.org/wiki/Software/systemd/):

  1) meson --prefix=/usr --buildtype=release -Dstrip=true build
  2) ninja -C build install
  3) systemctl daemon-reload
  4) systemctl enable nss-tlsd
  5) systemctl start nss-tlsd
  6) systemctl --user --global enable nss-tlsd
  7) systemctl --user start nss-tlsd
  8) ldconfig
  9) Add "tls" to the "hosts" entry in /etc/nsswitch.conf, before "dns" or
     anything else that contains "dns" in its name

This will enable a system nss-tlsd instance for all non-interactive processes
(which runs as an unprivileged user) and a private instance of nss-tlsd for each
user.

Choosing a DoH Server
=====================

By default, nss-tls performs encrypted name lookup over HTTPS through 1.1.1.1
(https://developers.cloudflare.com/1.1.1.1/dns-over-https/).

To use a different DNS over HTTPS (DoH) server, use the "resolver" build option:

  meson configure -Dresolver=dns9.quad9.net/dns-query

Performance
===========

DNS over HTTPS is much slower than DNS. Therefore, it is recommended to enable
cache of name lookup results. nscd(8) can do that.

To enable DNS cache on Debian (http://www.debian.org/) and derivatives:

  1) apt install unscd
  2) Set "enable-cache" for "hosts" to "yes" in /etc/nscd.conf
  3) systemctl enable unscd
  4) systemctl start unscd

Legal Information
=================

nss-tls is free and unencumbered software released under the terms of the GNU
Lesser General Public License as published by the Free Software Foundation;
either version 2.1 of the License, or (at your option) any later version
license.

nss-tls is not affiliated with 1.1.1.1, Cloudflare (https://www.cloudflare.com/)
or Quad9 (https://www.quad9.net/).

The ASCII art logo at the top was made using FIGlet (http://www.figlet.org/).


debian/changelog
nss-tls (1.1-1) unstable; urgency=medium

  [ Kan-Ru Chen ]
  * New upstream 1.1 release (Closes: #977303)
  * debian/control: bump Standards-Version to 4.5.1, no changes needed
  * debian/control: Add salsa Vcs-* info

  [ Dima Krasner ]
  * debian: make systemd.pc a build-time dependency
  * debian: give nss-tls priority over nss-resolve too

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Fri, 01 Jan 2021 14:34:22 +0900

nss-tls (1.0-1) unstable; urgency=medium

  [ Dima Krasner ]
  * debian: remove 0001-Add-simple-manpage-for-tlslookup.patch, upstreamed
  * debian: 8.8.8.8 wants dns.google in Host:
  * debian: 8.8.8.8 documentation recommends GET

  [ Kan-Ru Chen ]
  * New upstream version 1.0

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Wed, 15 Jul 2020 22:45:18 +0900

nss-tls (0~git20200523-2) unstable; urgency=medium

  * Correctly list license as LGPL-2.1+ in debian/copyright

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Mon, 13 Jul 2020 11:23:56 +0900

nss-tls (0~git20200523-1) unstable; urgency=medium

  * Initial release (Closes: #958710)

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Sat, 23 May 2020 22:01:40 +0900


ci.py
#!/usr/bin/python3
#
# This file is part of nss-tls.
#
# Copyright (C) 2018, 2019  Dima Krasner
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

from selenium import webdriver
from selenium.webdriver.firefox.firefox_binary import FirefoxBinary
from selenium.webdriver.firefox.options import Options
import os
import unittest
import sys

opts = Options()
if os.getenv("CI"):
    opts.headless = True

# use the native resolver (https://wiki.mozilla.org/Trusted_Recursive_Resolver)
prof = webdriver.FirefoxProfile()
prof.set_preference("network.trr.mode", 5)

driver = webdriver.Firefox(firefox_binary=FirefoxBinary(sys.argv[1]),
                           options=opts,
                           firefox_profile=prof)

class FirefoxTest(unittest.TestCase):
    def test_sanity(self):
        for d in sys.argv[2:]:
            url = "https://" + d
            driver.get(url)
            assert d in driver.current_url

if __name__ == "__main__":
    suite = unittest.TestLoader().loadTestsFromTestCase(FirefoxTest)
    result = unittest.TextTestRunner(verbosity=2).run(suite)
    driver.quit()
    sys.exit(not result.wasSuccessful())


nss-tls.c
/*
 * This file is part of nss-tls.
 *
 * Copyright (C) 2018, 2019, 2020  Dima Krasner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>
#include <sys/un.h>
#include <netinet/in.h>
#include <string.h>
#include <errno.h>
#include <netdb.h>
#include <nss.h>
#include <pthread.h>
#include <stdint.h>

#include "nss-tls.h"

static void cleanup(void *arg)
{
    close((int)(intptr_t)arg);
}

enum nss_status _nss_tls_gethostbyname2_r(const char *name,
                                          int af,
                                          struct hostent *ret,
                                          char *buf,
                                          size_t buflen,
                                          int *errnop,
                                          int *h_errnop)
{
    struct sockaddr_un sun = {.sun_family = AF_UNIX};
    struct timeval tv = {.tv_sec = NSS_TLS_TIMEOUT / 2, .tv_usec = 0};
    struct nss_tls_data *data = (struct nss_tls_data *)buf;
    const char *dir;
    ssize_t out, total;
    size_t len;
    int s, i, state;
    enum nss_status status = NSS_STATUS_TRYAGAIN;
    uint8_t count;

    if (buflen < sizeof(*data)) {
        *errnop = ERANGE;
        *h_errnop = NETDB_INTERNAL;
        return NSS_STATUS_TRYAGAIN;
    }

    *errnop = ENOENT;
    *h_errnop = NETDB_SUCCESS;

    const char *const socketpath = getenv("NSS_TLS_SOCKET");

    if (!socketpath || getuid() != geteuid()) {
        if (geteuid() == 0)
            strcpy(sun.sun_path, NSS_TLS_SOCKET_PATH);
        else {
            dir = getenv("XDG_RUNTIME_DIR");
            if (dir) {
                len = strlen(dir);
                if (len > sizeof(sun.sun_path) - sizeof("/"NSS_TLS_SOCKET_NAME))
                    return NSS_STATUS_TRYAGAIN;

                memcpy(sun.sun_path, dir, len);
                sun.sun_path[len] = '/';
                ++len;
                strncpy(sun.sun_path + len,
                        NSS_TLS_SOCKET_NAME,
                        sizeof(sun.sun_path) - len);
                sun.sun_path[sizeof(sun.sun_path) - 1] = '\0';
             } else
                strcpy(sun.sun_path, NSS_TLS_SOCKET_PATH);
        }
    } else {
        len = strlen (socketpath);
        if (len >= sizeof(sun.sun_path)) return EXIT_FAILURE;
        memcpy (sun.sun_path, socketpath, len + 1);
    }

    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &state);

    s = socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
    if (s < 0) {
        if (state != PTHREAD_CANCEL_DISABLE)
            pthread_setcancelstate(state, NULL);
        *errnop = EAGAIN;
        return NSS_STATUS_TRYAGAIN;
    }

    pthread_cleanup_push(cleanup, (void *)(intptr_t)s);

    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);

    if ((setsockopt(s, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv)) < 0) ||
        (setsockopt(s, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) < 0))
        goto pop;

    if (connect(s, (const struct sockaddr *)&sun, sizeof(sun)) < 0) {
        if (errno != ENOENT)
            goto pop;

        strcpy(sun.sun_path, NSS_TLS_SOCKET_PATH);
        if (connect(s, (const struct sockaddr *)&sun, sizeof(sun)) < 0)
            goto pop;
    }

    data->req.af = af;
    strncpy(data->req.name, name, sizeof(data->req.name));
    data->req.name[sizeof(data->req.name) - 1] = '\0';
    for (total = 0; total < sizeof(data->req); total += out) {
        out = send(s,
                   (unsigned char *)&data->req + total,
                   sizeof(data->req) - total,
                   MSG_NOSIGNAL);
        if (out <= 0)
            goto pop;
    }

    for (total = 0; total < sizeof(data->res); total += out) {
        out = recv(s,
                   (unsigned char *)&data->res + total,
                   sizeof(data->res) - total,
                   0);
        if (out < 0)
            goto pop;
        if (out == 0)
            break;
    }

    if (total == 0) {
        status = NSS_STATUS_NOTFOUND;
        goto pop;
    }

    if (total != sizeof(data->res))
        goto pop;

    if (data->res.cname[0]) {
        ret->h_name = data->res.cname;
        data->aliases[0] = data->req.name;
        data->aliases[1] = NULL;
    } else {
        ret->h_name = data->req.name;
        data->aliases[0] = NULL;
    }
    ret->h_aliases = data->aliases;
    ret->h_addrtype = af;
    data->addrs[0] = NULL;
    ret->h_addr_list = data->addrs;

    count = data->res.count;
    if (count == 0) {
        *h_errnop = HOST_NOT_FOUND;
        status = NSS_STATUS_NOTFOUND;
        goto pop;
    }
    if (count > NSS_TLS_ADDRS_MAX)
        count = NSS_TLS_ADDRS_MAX;

    switch (af) {
    case AF_INET:
        ret->h_length = sizeof(struct in_addr);
        break;

    case AF_INET6:
        ret->h_length = sizeof(struct in6_addr);
        break;

    default:
        status = NSS_STATUS_NOTFOUND;
        goto pop;
    }

    for (i = 0; i < count; ++i)
        data->addrs[i] = (char *)&data->res.addrs[i];
    data->addrs[i] = NULL;

    *errnop = 0;
    status = NSS_STATUS_SUCCESS;

pop:
    pthread_cleanup_pop(1);
    return status;
}


nss-tls.h
/*
 * This file is part of nss-tls.
 *
 * Copyright (C) 2018, 2019  Dima Krasner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <inttypes.h>
#include <netinet/in.h>
#include <arpa/nameser.h>

#define NSS_TLS_ADDRS_MAX 16

struct nss_tls_req {
    int af;
    char name[NS_MAXDNAME];
} __attribute__((packed));

struct nss_tls_res {
    uint8_t count;
    int64_t expiry;
    char cname[NS_MAXDNAME];
    union {
        struct in_addr in;
        struct in6_addr in6;
    } addrs[NSS_TLS_ADDRS_MAX];
} __attribute__((packed));

struct nss_tls_data {
    char *aliases[2];
    char *addrs[NSS_TLS_ADDRS_MAX + 1];
    struct nss_tls_req req;
    struct nss_tls_res res;
};


tlslookup.c
/*
 * This file is part of nss-tls.
 *
 * Copyright (C) 2019  Dima Krasner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <stdlib.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <netdb.h>
#include <nss.h>
#include <stdio.h>

#include "nss-tls.h"

extern enum nss_status _nss_tls_gethostbyname2_r(const char *name,
                                                 int af,
                                                 struct hostent *ret,
                                                 char *buf,
                                                 size_t buflen,
                                                 int *errnop,
                                                 int *h_errnop);

int main(int argc, char *argv[])
{
    struct nss_tls_data data;
    struct hostent ent;
    char buf[INET6_ADDRSTRLEN];
    int err, h_err, i, afs[] = {AF_INET, AF_INET6};
    uint16_t total = 0;
    uint8_t j;

    if (argc != 2) {
        fprintf(stderr,
                "Usage: tlslookup HOST\n"
                "Resolve the internet address of HOST.\n");
        return EXIT_FAILURE;
    }

    for (i = 0; i < sizeof(afs) / sizeof(afs[0]); ++i) {
        switch (_nss_tls_gethostbyname2_r(argv[1],
                                          afs[i],
                                          &ent,
                                          (char *)&data,
                                          sizeof(data),
                                          &err,
                                          &h_err)) {
        case NSS_STATUS_SUCCESS:
            break;

        case NSS_STATUS_NOTFOUND:
            continue;

        default:
            return EXIT_FAILURE;
        }

        for (j = 0; j < data.res.count; ++j) {
            if (!inet_ntop(afs[i], &data.res.addrs[j], buf, sizeof(buf)) ||
                (puts(buf) == EOF))
                return EXIT_FAILURE;
        }

        total += data.res.count;
    }

    if (total == 0)
        return EXIT_FAILURE;

    return EXIT_SUCCESS;
}


ci.sh
#!/bin/sh -xe

# This file is part of nss-tls.
#
# Copyright (C) 2018, 2019, 2020  Dima Krasner
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

DOMAINS="
    ipv4.google.com
    google.com
    youtube.com
    facebook.com
    baidu.com
    wikipedia.org
    taobao.com
    amazon.com
    twitter.com
    instagram.com
    reddit.com
    yandex.ru
    netflix.com
    aliexpress.com
    ebay.com
    bing.com
    github.com
"

# at least some Travis and AWS CodeBuild machines don't have an IPv6 route, so
# we only resolve these
IPV6_ONLY_DOMAINS="
    ipv6.google.com
"

meson --prefix=/usr --buildtype=release -Dstrip=true build
ninja -C build install

# make sure automatic DNS to DoH upgrade works
cat << EOF > /etc/resolv.conf
nameserver 1.1.1.1
nameserver 9.9.9.9
EOF
./build/nss-tlsd &
pid=$!
sleep 1
tlslookup google.com

# pick 3 random domains
domains=`echo $DOMAINS | tr ' ' \\\n | shuf | head -n 3`

for d in $domains
do
    tlslookup $d
done

kill -9 $pid
sleep 1

# make sure explicit choice of DoH servers works
echo -n > /etc/resolv.conf
meson configure build -Dresolvers=https://9.9.9.9/dns-query+random,https://dns.google/dns-query+random,https://1.1.1.1/dns-query+random
ninja -C build install

CC=clang meson --prefix=/usr -Db_sanitize=address build-asan
ninja -C build-asan nss-tlsd

ldconfig
echo "8.8.8.8 dns.google" >> /etc/hosts
cp -f /etc/nsswitch.conf /tmp/
sed 's/hosts:.*/hosts: files tls/' -i /etc/nsswitch.conf
./build-asan/nss-tlsd -r | tee /tmp/nss-tlsd.log &
pid=$!
sleep 1

for i in a b c
do
    for d in $domains $IPV6_ONLY_DOMAINS
    do
        valgrind --leak-check=full --track-fds=yes --error-exitcode=1 --errors-for-leak-kinds=all tlslookup $d
    done

    for d in $domains $IPV6_ONLY_DOMAINS
    do
        getent hosts $d
    done
done

# resolving the domain of a DoH server should always fail
tlslookup dns.google && exit 1

# resolving domains suffixed by the local domain should fail too and change of
# the local domain should take effect immediately
echo "search ci" >> /etc/resolv.conf
tlslookup google.com.ci && exit 1

# before 9169a0, the canonical name was an alias (instead of being the name,
# with the non-canonical domain being the alias), so handshakes failed if the
# certificate specified only the non-canonical name
for d in $domains
do
    wget -T 5 -t 2 -O /dev/null https://$d
done

# test against Firefox Nightly
[ -f firefox/firefox ] || wget -O- "https://download.mozilla.org/?product=firefox-nightly-latest-ssl&os=linux64&lang=en-US" | tar -xjf-
./ci.py firefox/firefox $domains

# before 963b0b, 8.8.8.8 responded with 400 if the dns= parameter contained URL
# unsafe characters
[ -n "`grep '^< HTTP/' /tmp/nss-tlsd.log | grep -v 200`" ] && exit 1

sed -i s/^resolvers=.*/resolvers=/ /etc/nss-tls.conf

# if resolving fails, we should try the next NSS module
echo "nameserver 185.228.168.168" > /etc/resolv.conf
kill -9 $pid
sleep 1
./build/nss-tlsd &
pid=$!
sleep 1
getent hosts google.com && exit 1
sed 's/hosts:.*/hosts: tls dns/' -i /etc/nsswitch.conf
getent hosts google.com

# if we have zero DoH servers, we should try the next NSS module
echo > /etc/resolv.conf
kill -9 $pid
sleep 1
./build/nss-tlsd &
pid=$!
sleep 1
echo "nameserver 9.9.9.9" > /etc/resolv.conf
tlslookup google.com && exit 1
getent hosts google.com

# if nss-tlsd is down, we should try the next NSS module
kill -9 $pid
sleep 1
tlslookup google.com && exit 1
getent hosts google.com

exit 0
